boa_engine/src/symbol.rs

```diff
@@ -260,46 +117,68 @@ struct Inner {
 }
 
 /// This represents a JavaScript symbol primitive.
-#[derive(Debug, Clone, Finalize)]
 pub struct JsSymbol {
-    inner: Rc<Inner>,
+    repr: Tagged<Inner>,
 }
 
+// SAFETY: `JsSymbol` uses `Arc` to do the reference counting, making this type thread-safe.
+unsafe impl Send for JsSymbol {}
+// SAFETY: `JsSymbol` uses `Arc` to do the reference counting, making this type thread-safe.
+unsafe impl Sync for JsSymbol {}
+
+impl Finalize for JsSymbol {}
+
 // Safety: JsSymbol does not contain any objects which needs to be traced,
 // so this is safe.
 unsafe impl Trace for JsSymbol {
     empty_trace!();
 }
 
+macro_rules! well_known_symbols {
+    ( $( $(#[$attr:meta])* ($name:ident, $variant:path) ),+$(,)? ) => {
+        $(
+            $(#[$attr])* pub(crate) const fn $name() -> JsSymbol {
+                JsSymbol {
+                    repr: Tagged::from_tag($variant.tag()),
+                }
+            }
+        )+
+    };
+}
+
 impl JsSymbol {
-    /// Create a new symbol.
+    /// Creates a new symbol.
+    ///
+    /// Returns `None` if the maximum number of possible symbols has been reached (`u64::MAX`).
     #[inline]
     #[must_use]
-    pub fn new(description: Option<JsString>) -> Self {
-        let hash = SYMBOL_HASH_COUNT.with(|count| {
-            let hash = count.get();
-            count.set(hash + 1);
-            hash
-        });
-
-        Self {
-            inner: Rc::new(Inner { hash, description }),
-        }
-    }
-
-    /// Create a new symbol with a specified hash and description.
-    #[inline]
-    fn with_hash(hash: u64, description: Option<JsString>) -> Self {
-        Self {
-            inner: Rc::new(Inner { hash, description }),
-        }
+    pub fn new(description: Option<JsString>) -> Option<Self> {
+        let hash = get_id()?;
+        let arc = Arc::new(Inner { hash, description });
+
+        Some(Self {
+            // SAFETY: Pointers returned by `Arc::into_raw` must be non-null.
+            repr: unsafe { Tagged::from_ptr(Arc::into_raw(arc).cast_mut()) },
+        })
```

Comment:
This looks interesting but not sure i understand it.
Could you explain why the JsSymbol holds a Tagged pointer over the Arc? Doesn't the Arc return a pointer anyway?

Reply:
> a JsSymbol was 8 bits in size, making it the same size as an usize

Isn't a `usize` 32bits on a 32bit machine, 64bits on a 64bit machine? i.e the pointer size? I've not known of a `usize` be only 1 byte. https://doc.rust-lang.org/std/primitive.usize.html

Reply:
Sorry! I mistyped. Edited my comment

Reply:
Also, I assume pointers of 64-bit in size in the comment, but also applies to 32-bit pointers because the tag itself is `usize`.